core/rio/hdt/src/main/java/org/eclipse/rdf4j/rio/hdt/HDTDictionarySectionPFC.java

```diff
@@ -181,4 +252,76 @@ protected void parse(InputStream is) throws IOException {
 		}
 		return arr;
 	}
+
+	/**
+	 * Encode a single block, removing the input strings to conserve memory
+	 * 
+	 * @param iter
+	 * @return encoded block
+	 */
+	private byte[] encodeBlock(Iterator<String> iter) {
+		byte[][] tmp = new byte[stringsBlock][];
+		int i = 0;
+
+		byte[] base = iter.next().getBytes(StandardCharsets.UTF_8);
+		iter.remove();
+		// copy base string into first entry, with a trailing NULL
+		tmp[i] = new byte[base.length + 1];
+		System.arraycopy(base, 0, tmp[i], 0, base.length);
+
+		byte[] prev = base;
+
+		// encode a block, with a maximum of strings per block (minus the first base string)
+		for (i = 1; iter.hasNext() && i < stringsBlock; i++) {
+			byte[] str = iter.next().getBytes(StandardCharsets.UTF_8);
+			iter.remove();
+
+			// remove the common part, and only store the offset and the suffix and trailing NULL
+			int common = mismatch(prev, str);
+			byte[] c = VByte.encode(common);
+			tmp[i] = new byte[c.length + str.length - common + 1];
+			System.arraycopy(c, 0, tmp[i], 0, c.length);
+			System.arraycopy(str, common, tmp[i], c.length, str.length - common);
```

Comment:
I'm glad you're using System.arraycopy, it's much better performance than a loop :)

Reply:
Indeed :-) One of the few Java-methods for low-level bit manipulations. Sometimes I do miss coding in c ...